File: app/src/main/java/netpro/keyTransmitter/Key.java
package netpro.keytransmitter;

import android.view.MotionEvent;
import android.view.View;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public abstract class Key implements Serializable {

    private static final long serialVersionUID = 5559614634094335358L;

    public enum Type {
        RELEASED("指を離したとき"),
        LONGPRESS("長押し"),
        PRESSING("押している間"),
        EMPTY("何もしない"),
        KNOB("つまみ");

        private String description;
        private static Map<String, Type> toTypeMap = new HashMap<>();

        static {
            for (Type type : values()) {
                toTypeMap.put(type.description, type);
            }
        }

        Type(String description) {
            this.description = description;
        }

        public String getDescription() {
            return description;
        }

        public static Type toType(String description) {
            return toTypeMap.get(description);
        }
    }

    protected int columnSpan;
    protected int rowSpan;
    protected String description;
    protected Type type;
    protected List<String> keyCodeList;

    /**
     * @param columnSpan
     * @param rowSpan
     * @param description
     */
    public Key(int columnSpan, int rowSpan, String description, Type type) {
        this.columnSpan = columnSpan;
        this.rowSpan = rowSpan;
        this.description = description;
        this.type = type;
        keyCodeList = new ArrayList<>();
    }

    public int getColumnSpan() {
        return columnSpan;
    }

    public int getRowSpan() {
        return rowSpan;
    }

    public String getDescription() {
        return description;
    }

    public Type getType() {
        return type;
    }

    public List<String> getKeyCodeList() {
        return keyCodeList;
    }

    public void setKeyCodeList(List<String> keyCodeList) {
        this.keyCodeList = keyCodeList;
    }

    public void addKeyCode(String keyCode) {
        keyCodeList.add(keyCode);
    }

    public void onActionDown(View view, MotionEvent motionEvent) {
    }

    public void onActionUp() {
    }

    public void onLongClick() {
    }

    public void onCancel() {
    }

    public void onMove(View view, MotionEvent motionEvent) {
    }

    protected void send(List<String> keyCodeList) {
        KeyTransmitter.INSTANCE.send(keyCodeList);
    }
}

File: app/src/main/java/netpro/keytransmitter/Key.kt
package netpro.keytransmitter

import android.view.MotionEvent
import android.view.View
import java.io.Serializable
import java.util.*

abstract class Key(columnSpan: Int, rowSpan: Int, description: String, type: Key.Type) : Serializable {

    enum class Type constructor(val description: String) {
        RELEASED("指を離したとき"),
        LONGPRESS("長押し"),
        PRESSING("押している間"),
        EMPTY("何もしない"),
        KNOB("つまみ"),
        FLICK("フリック");

        companion object {
            private val toTypeMap = HashMap<String, Type>()

            init {
                for (type in values()) {
                    toTypeMap.put(type.description, type)
                }
            }

            fun toType(description: String): Type? {
                return toTypeMap[description]
            }
        }
    }

    var columnSpan: Int = 0
        protected set
    var rowSpan: Int = 0
        protected set
    var description: String
        protected set
    var type: Type
        protected set
    var keyCodeList: MutableList<String>
        get

    init {
        this.columnSpan = columnSpan
        this.rowSpan = rowSpan
        this.description = description
        this.type = type
        keyCodeList = mutableListOf()
    }

    open fun addKeyCode(keyCode: String) {
        keyCodeList.add(keyCode)
    }

    open fun onActionDown(view: View, motionEvent: MotionEvent) {}

    open fun onActionUp(view: View, motionEvent: MotionEvent) {}

    open fun onLongClick() {}

    open fun onCancel() {}

    open fun onMove(view: View, motionEvent: MotionEvent) {}

    protected fun send(keyCodeList: List<String>) {
        KeyTransmitter.send(keyCodeList)
    }

    companion object {
        private const val serialVersionUID = 5559614634094335358L
    }
}